lib/fog/orchestration/util/recursive_hot_file_loader.rb

```diff
@@ -0,0 +1,227 @@
+            template_base_url = base_url_for_url(template_file)
+            raw_template = read_uri(template_file)
+            template = YAML.safe_load(raw_template, [Date])
+
+            Fog::Logger.debug("Template visited: #{@visited}")
+            @visited.add(template_file)
+          else
+            raise "template_file is not a string of the expected form"
+          end
+
+          get_file_contents(template, template_base_url)
+
+          template
+        end
+
+        # Traverse the template tree looking for get_file and type
+        #   and populating the @files attribute with their content.
+        #   Resource referenced by get_file and type are eventually
+        #   replaced with their absolute URI as done in heatclient
+        #   and shade.
+        #
+        def get_file_contents(from_data, base_url = nil)
+          Fog::Logger.debug("Processing #{from_data} with base_url #{base_url}")
+
+          # Recursively traverse the tree.
+          if recurse_if(from_data)
+            recurse_data = from_data.kind_of?(Hash) ? from_data.values : from_data
+            recurse_data.each do |value|
+              get_file_contents(value, base_url)
+            end
+          end
+
+          # I'm on a Hash, process it.
+          return unless from_data.kind_of?(Hash)
+          from_data.each do |key, value|
+            next if ignore_if(key, value)
+
+            # Resolve relative paths.
+            str_url = url_join(base_url, value)
+
+            next if @files.key?(str_url)
+
+            file_content = read_uri(str_url)
+
+            # get_file should not recurse hot templates.
+            if key == "type" && template_is_raw?(file_content) && !@visited.include?(str_url)
+              template = get_template_contents(str_url)
+              file_content = YAML.dump(template)
+            end
+
+            @files[str_url] = file_content
+            # replace the data value with the normalised absolute URL as required
+            #  by https://docs.openstack.org/heat/pike/template_guide/hot_spec.html#get-file
+            from_data[key] = str_url
+          end
+        end
+
+        private
+
+        # Retrive the content of a local or remote file.
+        #
+        # @param A local or remote uri.
+        #
+        # @raise ArgumentError if it's not a valid uri
+        #
+        # Protect open-uri from malign arguments like
+        #  - "|ls"
+        #  - multiline strings
+        def read_uri(uri_or_filename)
+          remote_schemes = %w[http https ftp]
+          Fog::Logger.debug("Opening #{uri_or_filename}")
+
+          begin
+            # Validate URI to protect from open-uri attacks.
+            url = URI(uri_or_filename)
+
+            # Remote schemes must contain an host.
+            raise ArgumentError if url.host.nil? && remote_schemes.include?(url.scheme)
+
+            # Encode URI with spaces.
+            uri_or_filename = URI.encode(URI.decode(URI(uri_or_filename).to_s))
```

Comment:
Do we need to address the obsolete warning?
you already have `url`, so the line could be `url_or_filename = URI.encode(URI.decode(url.to_s))`

Reply:
1- which obsolete warning?

2- Applied your suggestion and removed the 3rd `URI` call

Reply:
Replaced with gsub(/ /, "%20")